Carousel: use buttons, not links, for prev/next controls

patrickhlauke opened the pull request:
- expand the styles to neutralise border/background
- change docs page
- add extra unit test to check that links or buttons work as controls
- modify visual test to use buttons as well

Preview https://deploy-preview-32627--twbs-bootstrap.netlify.app/docs/5.0/components/carousel/

patrickhlauke committed:
Carousel: use buttons, not links, for prev/next

- expand the styles to neutralise border/background
- change docs page
- add extra unit test to check that links or buttons work as controls
- modify visual test to use buttons as well

patrickhlauke commented:
to clarify: using links still works, but it's not really semantically correct. not a breaking change, as the old markup still works. just better/cleaner to do it using `<button>`

patrickhlauke committed:
Merge branch 'main' into patrickhlauke-carousel

patrickhlauke commented:
this will need to be expanded if https://github.com/twbs/bootstrap/pull/32638 lands first

patrickhlauke committed:
Merge branch 'main' into patrickhlauke-carousel

patrickhlauke commented:
temporarily set to draft while i expand this following merging of  #32638

patrickhlauke committed:
Merge branch 'main' into patrickhlauke-carousel

patrickhlauke committed:
Use buttons for newly-merged extra carousel

patrickhlauke commented:
setting to draft to make sure the added carousel section (disabled touch) is added appropriately

patrickhlauke committed:
Merge branch 'main' into patrickhlauke-carousel

patrickhlauke commented:
ah, my mistake, i see this won't be affected by my planned upcoming change. as you were

XhmikosR commented:
There are more instances in examples?

Also, perhaps we should have callout/example mentioning that both work but for whatever reasons we suggest buttons?

patrickhlauke committed:
Edit carousel examples

- use buttons instead of links for prev/next
- remove `role="button"` from links that are actually links

XhmikosR commented on site/content/docs/5.0/examples/carousel-rtl/index.html:
```diff
@@ -59,7 +59,7 @@ <h1>عنوان المثال.</h1>
           <div class="carousel-caption">
             <h1>عنوان مثال آخر.</h1>
             <p>ثم تزامناً الفرنسي الإقتصادية دار. لكل عن الضغوط المتّبعة, أن حتى إختار المدن بالإنزال. عن الشمل بالفشل تلك, بل أراض أوزار بلديهما حول. دون لكون والتي ثم, كُلفة ويعزى استطاعوا أن لمّ. جُل بخطوط واحدة البشريةً.</p>
-            <p><a class="btn btn-lg btn-primary" href="#" role="button">أعرف أكثر</a></p>
```
isn't the role needed since we use `a`?

patrickhlauke replied:
they're not buttons though, are they? they're links going to some page, no?

XhmikosR replied:
Ah, true, those are captions. Might be worth splitting these to a separate PR?

patrickhlauke replied:
to clarify: if a link acts as a link (goes to another page or whatever), rather than doing something dynamic (opening a modal dialog, a dropdown, etc), then it should *not* have `role="button"`. and for these examples, I'm assuming these links are meant to signify links that act as links. whether the link is visually styled as a `btn` is usually irrelevant ("call-to-action links"), what counts is what the purpose of the control is.

patrickhlauke replied:
> Might be worth splitting these to a separate PR?

admittedly, i'm already getting lost with the various PRs i had going at the moment, and as this is just tweaking something in an example, in passing, not sure if it's really useful to split? your call though.

patrickhlauke replied:
ok reverted this change with the latest commit

XhmikosR replied:
Thanks! Don't forget to submit a separate PR please

patrickhlauke replied:
done https://github.com/twbs/bootstrap/pull/32788

patrickhlauke commented:
> There are more instances in examples?

Ah, good catch, those slipped me by. Addressed now.

> Also, perhaps we should have callout/example mentioning that both work but for whatever reasons we suggest buttons?

the "for whatever reason" part of your comment worries me a bit. i thought we (at least in core team) all understood/aknowledged that links are for linking (to another page, or a section within the current page like a skip anchor) and buttons are for dynamic in-page actions. 

i'll add a little mention/note about this though in a moment, sure.

patrickhlauke committed:
Merge branch 'main' into patrickhlauke-carousel

patrickhlauke committed:
Clarify that controls can be button or link

also adds code ticks to `id`

patrickhlauke committed:
Merge branch 'patrickhlauke-carousel' of https://github.com/twbs/bootstrap into patrickhlauke-carousel

patrickhlauke commented:
> i'll add a little mention/note about this though in a moment, sure.

done

patrickhlauke committed:
Merge branch 'main' into patrickhlauke-carousel

XhmikosR commented:
> the "for whatever reason" part of your comment worries me a bit. i thought we (at least in core team) all understood/aknowledged that links are for linking (to another page, or a section within the current page like a skip anchor) and buttons are for dynamic in-page actions.

I'm with you, but people might wonder, that's all I meant :)

patrickhlauke committed:
Merge branch 'main' into patrickhlauke-carousel

patrickhlauke committed:
Revert separate/unrelated change to CTA links

moved to separate PR

mdo commented on site/content/docs/5.0/components/carousel.md:
```diff
@@ -46,7 +46,7 @@ Here's a carousel with slides only. Note the presence of the `.d-block` and `.w-
 ### With controls
 
-Adding in the previous and next controls:
+Adding in the previous and next controls. We recommend using `<button>` elements, but you can also use `<a>` link elements–though in that case, you should also add `role="button"` attributes to the links.
```
Can make this short and sweet IMO.

```suggestion
Adding in the previous and next controls. We recommend using `<button>` elements, but you can also use `<a>` elements with `role="button"`.
```

patrickhlauke replied:
that might be misleading, as it may suggest our code only supports `<a>` with `role="button"`, while in effect it doesn't really check for `role="button"`?

patrickhlauke replied:
@mdo would you be ok with the more verbose, but possibly clearer, wording?

mdo replied:
I think brevity and best practice should win out—if folks use anchor elements, they really should use `role="button"`, yes? If that's not the case, then yeah we can use the more verbose.

patrickhlauke replied:
*shrug* fairy snuff :)

mdo reviewed (approved):
One suggested edit

patrickhlauke committed:
Merge branch 'main' into patrickhlauke-carousel

rohit2sharma95 committed:
Merge branch 'main' into patrickhlauke-carousel

patrickhlauke committed:
Update site/content/docs/5.0/components/carousel.md

Co-authored-by: Mark Otto <markd.otto@gmail.com>

patrickhlauke committed:
Merge branch 'main' into patrickhlauke-carousel

XhmikosR committed:
Merge branch 'main' into patrickhlauke-carousel

XhmikosR commented:
@patrickhlauke Is it intentional that the carousel controls sort of jump when clicked? Firefox 85 on Windows 10. Edge Chromium seems to be the same as before.

patrickhlauke committed:
Merge branch 'main' into patrickhlauke-carousel

patrickhlauke committed:
Explicitly set padding to 0 to prevent dipping/moving on active in Firefox

patrickhlauke commented:
> @patrickhlauke Is it intentional that the carousel controls sort of ~jump~move when clicked? Firefox 85 on Windows 10. Edge Chromium seems to be the same as before.

ah, good catch, hadn't even checked. no, wasn't intentional. forgot about Firefox's quirky behaviour of adding extra padding on `:active`. neutralised this now by explicitly setting `padding: 0` for the prev/next controls

XhmikosR commented on js/tests/unit/carousel.spec.js:
```diff
@@ -1154,16 +1154,40 @@ describe('Carousel', () => {
+        '    <div id="item2" class="carousel-item">item 2</div>',
+        '    <div class="carousel-item">item 3</div>',
+        '  </div>',
+        '  <a class="carousel-control-prev" href="#myCarousel" role="button" data-bs-slide="prev"></button>',
```
@patrickhlauke just to confirm, so in this branch we no longer have a JS test for `<div class="carousel-control-prev"` etc?

patrickhlauke replied:
correct. happy to add another test back in that tests explicitly for `<div role="button" tabindex="0" ...>` but that shouldn't really behave any different

XhmikosR replied:
Yeah, agreed. In JS we just look for the class.

XhmikosR commented:
Now, it looks good to me, thanks! One last comment and also I wonder, do we have an issue about this?

patrickhlauke commented:
> Now, it looks good to me, thanks! One last comment and also I wonder, do we have an issue about this?

we have this meta-issue where i reel off various things that should get fixed https://github.com/twbs/bootstrap/issues/31186 but no atomic issue about just this aspect i think. can do a trawl through open issues after this just to check though

patrickhlauke commented:
:+1: thanks for sanity checking @XhmikosR
